### ISSUE TITLE
How to send first stock order? [ Help wanted ]

### ISSUE THREAD

**Issue by Tsopic** [stoqey/ib#29]
Could someone from the team help me out with sending the first stock order? 

I'm getting Socket Error - 

```
ib.once(EventName.nextValidId, (orderId: number) => {
    const contract: Contract = {
      symbol: "AMZN",
      secType: SecType.STK,
    };

    const order: Order = {
      orderType: OrderType.LMT,
      action: OrderAction.BUY,
      lmtPrice: 4000,
      orderId,
    };

    ib.placeOrder(orderId, contract, order);
  });

ib.connect();
ib.reqIds();
```
The error I'm facing...

`
Error reading request.Message id 1.  Socket I/O error -  - code: 320 - reqId: 1
`
I was unable to determine the origin of where it fired off.

Would appreciate any help from the team. I'd also include the example in the readme.

**Comment by mfrener** [comment]
Hmm socket I/O error is strange. My first guess was a bug on placeOrder encoder, but it is not even even validating or processing data??.

It is an error coming from IB as a response to the placeOrder. 
Is there any hint on the gateway log maybe?
320 is "Server error when reading an API client request" (https://interactivebrokers.github.io/tws-api/message_codes.html).
Usually you will see some kind of _Error reading request:Message type 5. Unable to parse data. java.lang.NumberFormatException: For input string: "FUT". IB Error Code: 320_.  if the sent data is wrong. But this looks like issue on the socket connection rather than the data?

Just found out my paper account expired, so will need 24h to have it active again and to be able to test this on my own.

If you want to dig into in the meanwhile (we are still looking for more maintainers on this projects :P ): 
- First thing to look at would be gateway log. You should see the placeOrder command arriving and than hopefully some more info about what's going on. 
- If there is nothing suspicious on logs that indicates a socket issue, I would look at the placeOrder encoder.
https://github.com/stoqey/ib/blob/master/src/io/encoder.ts - placeOrder(id: number, contract: Contract, order: Order) is the TS counterpart to https://github.com/stoqey/ib/blob/master/ref/client/EClient.java public synchronized void placeOrder( int id, Contract contract, Order order). 
Easiest way to debug, is to run TS and Java clients parallel on step trough it.
To debug our lib:
* checkout the git repo
* code .
* copy/paste you code into api.spec.ts (at first test case, you don't want to wait for all other run first.. ;) )
* place a breakpoint on placeOrder call and hit F5.
To debug IB java client you can get the official API from IB and adapt their demo tools.
If you cannot debug the java part, doing a code review (cross-check what we do against what IB does) or just stepping trough the TS code to see what it actually does, can also help to locate the issue.

thanks for he PR btw ;)

**Comment by mfrener** [comment]
Ok, did a try on live (but with limit price 1$ :P ) and it actually is a bug on placeOrder, combined with missing attributes on the call itself. Please wait for merge of https://github.com/stoqey/ib/pull/30 or the fix locally if you want to pre-test.

Than, when running your code you will face the following log message on TWS / Gateway API that lead to solution:
- Placing API order: no account specified for multiple accounts user. This probably because if have multiple accounts and run gateway. If you have 1 account or run TWS and have an account selected, your not need account, otherwise you do.
- The contract description specified for AMZN is ambiguous;\nyou must specify the currency.
- Error validating request.-'bO' : cause - Missing order exchange.
- Error validating request.-'bO' : cause - The size value cannot be zero:
TWS APIO can be a bitch.. extend your code make it happy:
`const contract: Contract = {
        symbol: "AMZN",
        exchange: "SMART",
        currency: "USD",
        secType: SecType.STK,
      };
 const order: Order = {
        orderType: OrderType.LMT,
        action: OrderAction.BUY,
        lmtPrice: 1,
        orderId,
        totalQuantity: 1,
        account: <your_account_id>
      };
`
will prepare one more PR for readme, adding the placeOrder sample - or maybe you can after verifying that it works? ;)

**Comment by mfrener** [comment]
PR is merged, fixed in v1.0.7

**Comment by mfrener** [comment]
Readme is updated as well - closed this one.
